### ISSUE TITLE
Inconsistent Skybox Rendering in Pre/Post Translucent/Opaque Renderables Hooks

### ISSUE THREAD

**Issue by A1steaksa** [Facepunch/garrysmod-issues#5780]
### Details
From what I've been able to figure out through testing, there appears to be problem(s) with the rendering of skybox elements in the following hooks:
* `PreDrawOpaqueRenderables`
* `PostDrawOpaqueRenderables`
* `PreDrawTranslucentRenderables`
* `PostDrawTranslucentRenderables`

I've been trying to figure out why the entities I'm drawing in my project's `PostDrawTranslucentRenderables` hook were drawing both in the world and in the skybox.  It turns out that the behavior around these four hooks and the skybox is strange.

I've been using `render.PushRenderTarget( nil )` as a method to render to the screen based on some advice I was given, so that is included in the testing I've done.  I'm not 100% sure if that's an officially sanctioned and supported thing to do, but it's very useful to be able to do.  It seems that pushing that `nil` Render Target has a significant impact on the behavior of these hooks, as shown below.  Even without doing that, though, there is still some strange behavior being shown from my testing.

These are the results of my testing:
Hook | Pushed nil Render Target? | Is Any Skybox Being Drawn? | Entity draws in the skybox? | Entity draws in the world? | Notes
-- | -- | -- | -- | -- | --
PreDrawOpaqueRenderables | No | Yes | No | No | Not Drawing in Skybox
PreDrawOpaqueRenderables | No | No | No | Yes | Normal Behavior
PostDrawOpaqueRenderables | No | Yes | No | No | Not Drawing in Skybox
PostDrawOpaqueRenderables | No | No | No | Yes | Normal Behavior
PreDrawTranslucentRenderables | No | Yes | No | No | Not Drawing in Skybox
PreDrawTranslucentRenderables | No | No | No | Yes | Normal Behavior
PostDrawTranslucentRenderables | No | Yes | No | No | Not Drawing in Skybox
PostDrawTranslucentRenderables | No | No | No | Yes | Normal Behavior
PreDrawOpaqueRenderables | Yes | Yes | Yes | No | Normal Behavior
PreDrawOpaqueRenderables | Yes | No | No | Yes | Normal Behavior
PostDrawOpaqueRenderables | Yes | Yes | Yes | No | Normal Behavior
PostDrawOpaqueRenderables | Yes | No | No | Yes | Normal Behavior
PreDrawTranslucentRenderables | Yes | Yes | No | No | Not Drawing in Skybox
PreDrawTranslucentRenderables | Yes | No | Yes | Yes | Drawing in Both
PostDrawTranslucentRenderables | Yes | Yes | No | No | Not Drawing in Skybox
PostDrawTranslucentRenderables | Yes | No | Yes | Yes | Drawing in Both

### Reproduction

Testing was done on the `x86-64 - Chromium + 64-bit binaries` branch on `gm_construct` and `rp_bangclaw`.
I also did some testing in the `main`, `dev`, and `pre-release` branches but I didn't see any change in behavior between them.  The testing between branches was not thorough enough to include here.

Using the following code, uncomment one of the `DrawTest()` function calls and comment/uncomment the `render.PushRenderTarget( nil )` and `render.PopRenderTarget()` within `DrawText()`'s definition to iterate through the test cases.

```lua
local clientsideModel = ClientsideModel( "models/props_phx/construct/windows/window_curve360x2.mdl" )
clientsideModel:SetPos( Vector( 0, 0, 0 ) )
clientsideModel:SetAngles( Angle( 0, 0, 0 ) )
clientsideModel:SetNoDraw( true )
clientsideModel:Spawn()

local testPos = Vector( 10, 25, -70 )
local testAng = Angle( 10, 45, 90 )

local function DrawTest()
    --render.PushRenderTarget( nil )
        render.Model( { model = clientsideModel:GetModel(), pos = testPos, angle = testAng }, clientsideModel )
    --render.PopRenderTarget()
end

hook.Add( "PreDrawOpaqueRenderables", "SkyboxErrorTestPreOpaque", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
    local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
    if isAnySkyboxBeingDrawn then
        --DrawTest() -- Should draw in skybox
    else
        --DrawTest() -- Should draw in world
    end
end)

hook.Add( "PostDrawOpaqueRenderables", "SkyboxErrorTestPostOpaque", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
    local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
    if isAnySkyboxBeingDrawn then 
        --DrawTest() -- Should draw in skybox
    else
        --DrawTest() -- Should draw in world
    end
end)

hook.Add( "PreDrawTranslucentRenderables", "SkyboxErrorTestPreTranslucent", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
    local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
    if isAnySkyboxBeingDrawn then 
        --DrawTest() -- Should draw in skybox
    else
        --DrawTest() -- Should draw in world
    end
end)

hook.Add( "PostDrawTranslucentRenderables", "SkyboxErrorTestPostTranslucent", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
    local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
    if isAnySkyboxBeingDrawn then 
        --DrawTest() -- Should draw in skybox
    else
        --DrawTest() -- Should draw in world
    end
end)
```

**Comment by robotboy655** [comment]
So, from my testing using a modified version of your script, it would appear that drawing during `isDrawingSkybox == true` specifically would still draw the model in not skybox?

I am noticing that with 3D skybox disabled (`r_3dsky 0`) the value of `isAnySkyboxBeingDrawn` appears to be inverted during the 2 OpaqueRenderables hooks. That's the only inconsistency I could find.

That being said, do not expect to be able to render things in 3D skybox on maps that do not have it.

I have not observed any differences with pushing the rendertarget (`render.PushRenderTarget( nil )` would push the main render target to the stack, which is in all likelyhood a waste of resources at best, since during those hooks the active rendertarget should be the main one anyway)

All my testing was done on `dev` as you have claimed to observe no differences between `dev` and `x86-64`.

https://github.com/Facepunch/garrysmod-issues/assets/3299036/f56ebb9c-650b-4b21-a11f-013ab01c231a


Here's the script I used:
```
	local clientsideModel = ClientsideModel( "models/props_phx/construct/windows/window_curve360x2.mdl" )
	clientsideModel:SetPos( Vector( 0, 0, 0 ) )
	clientsideModel:SetAngles( Angle( 0, 0, 0 ) )
	clientsideModel:SetNoDraw( true )
	clientsideModel:Spawn()
	
	local testPos = Vector( -2755.627197, 2921.753418, 12584.780273 )
	local testAng = Angle( 10, 45, 90 )
	
	local sky = false;
	local hookId = 1
	local hooks = {
		"Pre  DrawOpaqueRenderables",
		"PostDrawOpaqueRenderables",
		"Pre  DrawTranslucentRenderables",
		"PostDrawTranslucentRenderables",
	}
	local hookColors = {
		Color( 255, 0, 0),
		Color( 255, 255, 0),
		Color( 0, 255, 0),
		Color( 0, 0, 255)
	}

	hook.Add( "HUDPaint", "test", function ()
		draw.SimpleText( "SKY: " .. tostring( sky ), "DermaLarge", 100, 130, color_white )
		draw.SimpleText( "Hook: " .. hooks[ hookId ], "DermaLarge", 100, 100, color_white )

		sky = CurTime() % 2 < 1
		hookId = math.floor( CurTime() % 8 /2 ) + 1
	end)

	local function DrawTest()
		--render.PushRenderTarget( nil )
		render.SetColorModulation( hookColors[hookId]:Unpack() )
		render.Model( { model = clientsideModel:GetModel(), pos = testPos, angle = testAng }, clientsideModel )
		--render.PopRenderTarget()
	end
	
	hook.Add( "PreDrawOpaqueRenderables", "SkyboxErrorTestPreOpaque", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
		local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
		if ( hookId != 1 ) then return end
		
		if isAnySkyboxBeingDrawn then 
			if ( sky ) then DrawTest() end -- Should draw in skybox
		else
			if ( !sky ) then DrawTest() end -- Should draw in world
		end
	end)
	
	hook.Add( "PostDrawOpaqueRenderables", "SkyboxErrorTestPostOpaque", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
		local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
		if ( hookId != 2 ) then return end

		if isAnySkyboxBeingDrawn then 
			if ( sky ) then DrawTest() end -- Should draw in skybox
		else
			if ( !sky ) then DrawTest() end -- Should draw in world
		end
	end)
	
	hook.Add( "PreDrawTranslucentRenderables", "SkyboxErrorTestPreTranslucent", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
		local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
		if ( hookId != 3 ) then return end
		if isAnySkyboxBeingDrawn then 
			if ( sky ) then DrawTest() end -- Should draw in skybox
		else
			if ( !sky ) then DrawTest() end -- Should draw in world
		end
	end)
	
	hook.Add( "PostDrawTranslucentRenderables", "SkyboxErrorTestPostTranslucent", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
		local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
		if ( hookId != 4 ) then return end
		if isAnySkyboxBeingDrawn then 
			if ( sky ) then DrawTest() end -- Should draw in skybox
		else
			if ( !sky ) then DrawTest() end -- Should draw in world
		end
	end)
```

**Comment by robotboy655** [comment]
Infact the apparent reversed arguments are a side effect of my own test script, without a 3D skybox the 2D skybox is always drawn during the opaque entity phase of the "main view", as opposed to having 2 separate render passes for the 3d skybox.

In either case, you should only ever worry about the 3d skybox argument.

**Comment by A1steaksa** [comment]
Thanks for taking the time to give it a look.

To make things more simple, here's some footage of one of the specific types of issues I've been seeing:
When drawing in an opaque renderable hook AND while pushing a `nil` Render Target, AND while drawing the skybox, rendered models appear in the skybox, but ONLY when standing in one X, Y quadrant of the map relative to the origin.

The map I'm using is `rp_bangclaw` but the same happens on `gm_construct` but it's harder to see there.

In the following footage, I'm stepping around the origin point of the map, seen here:
![image](https://github.com/Facepunch/garrysmod-issues/assets/8369707/105f756c-e2c5-430d-a4de-e2883ce8c01f)

https://github.com/Facepunch/garrysmod-issues/assets/8369707/9d9a3945-fd80-4cab-a944-c8228ff827f1

The code I'm using is
```lua
local clientsideModel = ClientsideModel( "models/props_phx/construct/windows/window_curve360x2.mdl" )
clientsideModel:SetPos( Vector( 0, 0, 0 ) )
clientsideModel:SetAngles( Angle( 0, 0, 0 ) )
clientsideModel:SetNoDraw( true )
clientsideModel:Spawn()

local testPos = Vector( -151.894577, -1196.434082, 150 )
local testAng = Angle( 10, 45, 90 )

hook.Add( "Think", "SkyboxTestingOrigin", function()
    debugoverlay.Sphere( Vector( 0, 0, 0 ), 20, 0.1, Color( 255, 255, 255 ), true )
    debugoverlay.Text( Vector( 0, 0, 0 ), "Origin (0, 0, 0)", 0.1, false )
    debugoverlay.Axis( Vector( 0, 0, 0 ), Angle( 0, 0, 0 ), 1000, 0.1, true )
    debugoverlay.Axis( Vector( 0, 0, 0 ), Angle( 180, 90, 0 ), 1000, 0.1, true )
end )

hook.Add( "PostDrawOpaqueRenderables", "SkyboxTesting", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
    local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
    
    render.PushRenderTarget( nil )

    if isAnySkyboxBeingDrawn then
        render.Model( { model = clientsideModel:GetModel(), pos = testPos, angle = testAng }, clientsideModel )
    end

   render.PopRenderTarget()
end)
```

Hopefully a single, specific example will make it easier to recreate the issue(s).

**Comment by robotboy655** [comment]
I see. So the issue is in fact entirely revolving around calling `render.PushRenderTarget( nil )` during skybox rendering?
What exactly are you pushing the RT for?

From what I can tell, the reason "it only renders while looking at certain location" (and the position/angle changes view your view) is because what is happening is that you have water in your viewport during that time, so the PostDrawOpaqueRenderables hook gets called for the water reflection render pass (including additional 3d skybox render pass), but with the active render target set to the water reflection one. So you reset it to the main framebuffer one with your `render.PushRenderTarget()` call and end up rendering to the wrong RT.

To me this feels like an "deliberately affecting engine rendering and then complaining that it is affected" sort-of issue.

I really don't understand what the purpose of the `render.PushRenderTarget()` call is (I don't think even you yourself do), but if you really MUST have it for some inexplicable reason, I imagine this could be the solution:
```lua
hook.Add( "PostDrawOpaqueRenderables", "SkyboxTesting", function( isDrawingDepth, isDrawingSkybox, isDrawing3dSkybox )
    local isAnySkyboxBeingDrawn = isDrawingSkybox or isDrawing3dSkybox
    
	local RT = render.GetRenderTarget()
    if ( RT == nil ) then render.PushRenderTarget() end

    if isAnySkyboxBeingDrawn then
		render.Model( { model = clientsideModel:GetModel(), pos = testPos, angle = testAng }, clientsideModel )
    end

	if ( RT == nil ) then render.PopRenderTarget() end
end)
```

I would also, again, strongly advise on only using the 3rd argument if you wish to (not) render stuff in the 3D skybox.

**Comment by A1steaksa** [comment]
Oh!  I understand what's happening now.  Your explanation of it being part of the reflection render pass made the whole thing make sense.

> I don't think even you yourself do

You have so little faith in me!

> So the issue is in fact entirely revolving around calling render.PushRenderTarget( nil ) during skybox rendering?
> What exactly are you pushing the RT for?  

This issue came up while I was making a system to render arbitrary entities at a lower resolution and composite them back into the scene.  I pushed the low resolution render target, then drew all of the entities to it but I also needed to set up a Stencil mask for those entities in the full resolution scene so I was pushing the `nil` render target and then rendering the entity there to set up the Stencil mask because I was pushing the low resolution render target from outside that function.  I elected to push `nil`, draw, pop instead of pop, draw, push `the low resolution render target`.  They seemed like equivalent options at the time and pushing `nil` meant I didn't have to pass in another argument to the function.

> the PostDrawOpaqueRenderables hook gets called for the water reflection render pass (including additional 3d skybox render pass), but with the active render target set to the water reflection one. So you reset it to the main framebuffer one with your render.PushRenderTarget() call and end up rendering to the wrong RT.

This explanation was extremely helpful for making me understand what the problem was and what's happening!

> To me this feels like a "deliberately affecting engine rendering and then complaining that it is affected" sort-of issue.

It 100% is.  The problem was that I didn't **realize** I was affecting engine rendering.  I've seen the render hook order page, but it doesn't really explain how/why each of the steps is carried out and what the outcome is so it's hard to know exactly what's happening in each step and how your actions might influence/interfere with them.

Thanks again for taking the time to look into this for me.  I really appreciate it!
